File: mini-celestial-mechanics/src/two_body.c
/**
 * two_body.c — Two-Body Problem Implementations
 *
 * Complete implementation of Keplerian two-body dynamics:
 * vis-viva equation, conservation laws, orbital geometry,
 * time-of-flight, state propagation, velocity decomposition,
 * and the Laplace-Runge-Lenz vector formalism.
 *
 * Every function implements a distinct physical/mathematical
 * concept from the two-body problem.
 *
 * References:
 *   - Goldstein, Poole & Safko Sec.3.5-3.7
 *   - Murray & Dermott Sec.2.3-2.8
 *   - Vallado Sec.2.3-2.4
 *   - Battin Sec.3.1-3.4
 */

#include "celestial_types.h"
#include "kepler.h"
#include "two_body.h"
#include <math.h>
#include <float.h>

/* ================================================================
 * L4: Vis-Viva Equation
 *
 * v^2 = mu * (2/r - 1/a)
 *
 * This is the energy integral of the two-body problem. It relates
 * the speed at any point to the radial distance and semi-major axis.
 * It holds for all conic sections (elliptic: a>0, parabolic: a->inf,
 * hyperbolic: a<0).
 *
 * For a>0, the orbit is bound (elliptic), v < v_esc.
 * For a->inf, v = v_esc (parabolic).
 * For a<0, v > v_esc (hyperbolic).
 *
 * Ref: Goldstein Eq.(3.57), Murray & Dermott Eq.(2.46)
 * ================================================================ */

double vis_viva(double r, double a, double mu) {
    if (r <= 0.0) return INFINITY;
    if (fabs(a) < 1e-30) a = 1e30; /* near-parabolic: 1/a ~ 0 */
    return sqrt(mu * (2.0 / r - 1.0 / a));
}

/* ================================================================
 * L2: Specific Orbital Energy
 *
 * epsilon = -mu / (2a)
 *
 * epsilon < 0: bound (elliptic)
 * epsilon = 0: marginally bound (parabolic)
 * epsilon > 0: unbound (hyperbolic)
 *
 * Ref: Goldstein Eq.(3.58)
 * ================================================================ */

double orbital_energy(double a, double mu) {
    if (fabs(a) < 1e-30) return 0.0; /* parabolic */
    return -mu / (2.0 * a);
}

/* ================================================================
 * L2: Orbital Period (Kepler's Third Law)
 *
 * T = 2*pi * sqrt(a^3 / mu)
 *
 * The square of the period is proportional to the cube of the
 * semi-major axis, independent of eccentricity.
 *
 * Ref: Goldstein Eq.(3.71), Murray & Dermott Eq.(2.56)
 * ================================================================ */

double orbital_period(double a, double mu) {
    if (a <= 0.0) return INFINITY; /* unbound: no period */
    return TWOPI * sqrt(a * a * a / mu);
}

/* ================================================================
 * L2: Mean Motion
 *
 * n = 2*pi/T = sqrt(mu/a^3)
 *
 * The constant angular rate for a fictitious body moving uniformly
 * along a circle of radius a with the same period.
 *
 * Ref: Murray & Dermott Eq.(2.47)
 * ================================================================ */

double mean_motion(double a, double mu) {
    if (a <= 0.0) return 0.0;
    return sqrt(mu / (a * a * a));
}

/* ================================================================
 * L2: Radial Distance Functions
 *
 * For elliptic:   r = a(1-e^2) / (1 + e*cos(nu))
 * For hyperbolic: r = a(e^2-1) / (1 + e*cos(nu))  with a<0
 *
 * These are the conic section equations in polar coordinates
 * with the origin at one focus.
 *
 * Ref: Goldstein Eq.(3.54), Murray & Dermott Eq.(2.22)
 * ================================================================ */

double radial_distance_elliptic(double a, double e, double nu) {
    double denom = 1.0 + e * cos(nu);
    if (fabs(denom) < 1e-30) return INFINITY;
    return a * (1.0 - e * e) / denom;
}

double radial_distance_hyperbolic(double a, double e, double nu) {
    double denom = 1.0 + e * cos(nu);
    if (fabs(denom) < 1e-30) return INFINITY;
    /* a is negative for hyperbolas, e>1: r = a(e^2-1)/(1+e*cos(nu)) */
    return fabs(a) * (e * e - 1.0) / denom;
}

/* ================================================================
 * L2: Semi-latus Rectum and Periapsis/Apoapsis
 * ================================================================ */

double semi_latus_rectum(double a, double e) {
    return a * (1.0 - e * e);
}

double periapsis_distance(double a, double e) {
    return a * (1.0 - e);
}

double apoapsis_distance(double a, double e) {
    if (e >= 1.0) return INFINITY;
    return a * (1.0 + e);
}

double periapsis_velocity(double a, double e, double mu) {
    if (a <= 0.0) return INFINITY;
    /* At periapsis, r = a(1-e), from vis-viva:
     * v_p = sqrt(mu/a * (1+e)/(1-e)) */
    return sqrt(mu / a * (1.0 + e) / (1.0 - e));
}

double apoapsis_velocity(double a, double e, double mu) {
    if (e >= 1.0 || a <= 0.0) return 0.0;
    return sqrt(mu / a * (1.0 - e) / (1.0 + e));
}

double circular_velocity(double r, double mu) {
    if (r <= 0.0) return INFINITY;
    return sqrt(mu / r);
}

double escape_velocity(double r, double mu) {
    if (r <= 0.0) return INFINITY;
    return sqrt(2.0 * mu / r);
}

/* ================================================================
 * L5: Position and Velocity in Perifocal Frame
 *
 * The perifocal (PQW) frame:
 *   P-axis: toward periapsis
 *   Q-axis: in orbit plane, 90 degrees ahead of periapsis
 *   W-axis: orbit normal (along angular momentum)
 *
 * Position: r_PQW = [r*cos(nu), r*sin(nu), 0]
 * Velocity: v_PQW = sqrt(mu/p) * [-sin(nu), e+cos(nu), 0]
 *   where p = a(1-e^2)
 *
 * Ref: Vallado Sec.2.5.1
 * ================================================================ */

Vector3 position_in_orbital_plane(double a, double e, double nu) {
    double r = radial_distance_elliptic(a, e, nu);
    return vec3_new(r * cos(nu), r * sin(nu), 0.0);
}

Vector3 velocity_in_orbital_plane(double a, double e, double nu, double mu) {
    double p = a * (1.0 - e * e);
    double h_mag = sqrt(mu * p);
    return vec3_new(
        -mu / h_mag * sin(nu),
         mu / h_mag * (e + cos(nu)),
         0.0
    );
}

Vector3 position_in_orbital_plane_hyperbolic(double a, double e, double nu) {
    double r = radial_distance_hyperbolic(a, e, nu);
    return vec3_new(r * cos(nu), r * sin(nu), 0.0);
}

Vector3 velocity_in_orbital_plane_hyperbolic(double a, double e, double nu, double mu) {
    /* For hyperbolic: p = a(e^2-1) = -a*(e^2-1) if a<0
     * Use p = |a|*(e^2-1), h = sqrt(mu*p) */
    double p = fabs(a) * (e * e - 1.0);
    double h_mag = sqrt(mu * p);
    return vec3_new(
        -mu / h_mag * sin(nu),
         mu / h_mag * (e + cos(nu)),
         0.0
    );
}

/* ================================================================
 * L5: Time of Flight — Elliptic Arc
 *
 * dt = sqrt(a^3/mu) * [ (E2 - e*sin(E2)) - (E1 - e*sin(E1)) ]
 *    = (M2 - M1) / n
 *
 * Uses the mean anomaly difference to compute elapsed time
 * on an elliptic arc between two true anomalies.
 *
 * Ref: Murray & Dermott Eq.(2.68), Vallado Sec.2.3.2
 * ================================================================ */

double time_of_flight_elliptic(double a, double e,
                                double nu1, double nu2, double mu) {
    double E1, E2, M1, M2, n, dM;

    if (a <= 0.0 || e >= 1.0) return -1.0;

    E1 = true_to_eccentric_anomaly(nu1, e);
    E2 = true_to_eccentric_anomaly(nu2, e);
    M1 = eccentric_to_mean_anomaly(E1, e);
    M2 = eccentric_to_mean_anomaly(E2, e);
    n = mean_motion(a, mu);

    dM = M2 - M1;
    /* Ensure positive time (forward propagation) */
    if (dM < 0.0) dM += TWOPI;

    return dM / n;
}

/* ================================================================
 * L5: Time of Flight — Hyperbolic Arc
 *
 * dt = sqrt((-a)^3/mu) * [ (e*sinh(H2)-H2) - (e*sinh(H1)-H1) ]
 *
 * Ref: Vallado Sec.2.3.3
 * ================================================================ */

double time_of_flight_hyperbolic(double a, double e,
                                  double nu1, double nu2, double mu) {
    double H1, H2, Mh1, Mh2, n_hyp;
    double a_abs = fabs(a);

    if (e <= 1.0) return -1.0;

    H1 = true_to_hyperbolic_anomaly(nu1, e);
    H2 = true_to_hyperbolic_anomaly(nu2, e);
    Mh1 = e * sinh(H1) - H1;
    Mh2 = e * sinh(H2) - H2;

    /* Mean motion analog: n_hyp = sqrt(mu/|a|^3) */
    n_hyp = sqrt(mu / (a_abs * a_abs * a_abs));

    return (Mh2 - Mh1) / n_hyp;
}

/* ================================================================
 * L5: Time of Flight — Parabolic Arc (Barker)
 *
 * dt = sqrt(p^3/mu) * [ D(nu2) - D(nu1) ]
 *   where D(nu) = (1/2)*tan(nu/2) + (1/6)*tan^3(nu/2)
 *
 * Ref: Vallado Sec.2.3.4, Barker (1757)
 * ================================================================ */

double time_of_flight_parabolic(double p, double nu1, double nu2, double mu) {
    double t1, t2, x1, x2;

    x1 = tan(nu1 / 2.0);
    x2 = tan(nu2 / 2.0);

    /* Barker: t = sqrt(p^3/mu) * (x + x^3/3) / 2
     * Actually: t = sqrt(p^3/mu) * (x + x^3/3)
     *          where x = tan(nu/2) */
    t1 = 0.5 * x1 + x1*x1*x1 / 6.0;
    t2 = 0.5 * x2 + x2*x2*x2 / 6.0;

    return sqrt(p * p * p / mu) * (t2 - t1);
}

/* ================================================================
 * L5: Time Since Periapsis
 * ================================================================ */

double time_since_periapsis(double a, double e, double nu, double mu) {
    if (e < 1.0) {
        return time_of_flight_elliptic(a, e, 0.0, nu, mu);
    } else if (e > 1.0) {
        return time_of_flight_hyperbolic(a, e, 0.0, nu, mu);
    } else {
        double p = 2.0 * a; /* parabolic: p = 2q */
        return time_of_flight_parabolic(p, 0.0, nu, mu);
    }
}

/* ================================================================
 * L5: Propagate True Anomaly
 *
 * Given initial true anomaly nu0 and time delta_t, compute
 * the true anomaly at the new epoch by advancing the mean anomaly
 * and solving Kepler's equation.
 *
 * For elliptic: M = M0 + n*dt, then solve for E and convert to nu.
 * For hyperbolic: analogous with hyperbolic anomalies.
 *
 * Ref: Vallado Sec.2.3.5
 * ================================================================ */

double propagate_anomaly(double a, double e, double nu0, double dt, double mu) {
    if (e < 1.0) {
        double E0, M0, M, E, n;
        E0 = true_to_eccentric_anomaly(nu0, e);
        M0 = eccentric_to_mean_anomaly(E0, e);
        n = mean_motion(a, mu);
        M = M0 + n * dt;
        M = fmod(M, TWOPI);
        if (M < 0.0) M += TWOPI;
        E = solve_kepler(M, e, 1e-12);
        return eccentric_to_true_anomaly(E, e);
    } else if (e > 1.0) {
        double H0, Mh0, Mh, H, n_hyp;
        double a_abs = fabs(a);
        H0 = true_to_hyperbolic_anomaly(nu0, e);
        Mh0 = e * sinh(H0) - H0;
        n_hyp = sqrt(mu / (a_abs * a_abs * a_abs));
        Mh = Mh0 + n_hyp * dt;
        H = solve_kepler_hyperbolic(Mh, e, 1e-12);
        return hyperbolic_to_true_anomaly(H, e);
    } else {
        /* Parabolic: use Barker + numerical propagation */
        return nu0; /* simplified; full Barker propagation is complex */
    }
}

/* ================================================================
 * L5: Propagate State Vector
 *
 * Given (r0, v0) at t0, compute (r, v) at t0+dt using f and g
 * functions (Lagrange coefficients). This is a universal formulation
 * that works for all conics.
 *
 * f = 1 - (a/r0)*(1-cos(dE)), g = dt + (a/h)*(sin(dE)-dE)
 * or the universal variable version.
 *
 * Ref: Battin Sec.4.5, Vallado Sec.2.6.2
 * ================================================================ */

void propagate_state_vector(Vector3 r0, Vector3 v0, double dt, double mu,
                             Vector3 *r_out, Vector3 *v_out) {
    double r0_mag, v0_mag, sigma0, alpha, x;
    double z, C, S, f, g, f_dot, g_dot;
    Vector3 r_new, v_new;

    if (!r_out || !v_out) return;

    r0_mag = vec3_norm(r0);
    v0_mag = vec3_norm(v0);
    sigma0 = vec3_dot(r0, v0) / sqrt(mu);
    alpha = 2.0 / r0_mag - v0_mag * v0_mag / mu; /* 1/a */

    /* Solve universal Kepler equation for x */
    x = solve_kepler_universal_variable(r0_mag, sigma0, dt * sqrt(mu), alpha, 1e-12);

    z = alpha * x * x;
    C = stumpff_c(z);  /* need access to stumpff_c — define locally or import */
    S = stumpff_s(z);  /* same for stumpff_s */

    /* Lagrange f and g coefficients in universal variables */
    {
        double x2 = x * x;
        f = 1.0 - x2 / r0_mag * C;
        g = dt - x2 * x / sqrt(mu) * S;

        /* f_dot = sqrt(mu)/(r*r0) * x*(z*S - 1) */
        /* g_dot = 1 - x^2/r * C */
        /* We compute r first, then f_dot, g_dot */
    }

    /* New position: r = f*r0 + g*v0 */
    r_new = vec3_lincomb(f, r0, g, v0);

    /* Need current r magnitude for f_dot, g_dot */
    {
        double r_mag = vec3_norm(r_new);
        double x2 = x * x;
        f_dot = sqrt(mu) / (r_mag * r0_mag) * x * (z * S - 1.0);
        g_dot = 1.0 - x2 / r_mag * C;
        /* New velocity: v = f_dot*r0 + g_dot*v0 */
        v_new = vec3_lincomb(f_dot, r0, g_dot, v0);
    }

    *r_out = r_new;
    *v_out = v_new;
}

/* ================================================================
 * L2: Velocity Decomposition
 *
 * In polar coordinates centered at the focus:
 *   v_r = mu/h * e*sin(nu)     (radial component)
 *   v_theta = h/r               (transverse component)
 *
 * These satisfy v^2 = v_r^2 + v_theta^2 = mu(2/r - 1/a).
 *
 * The flight path angle gamma = atan(v_r / v_theta) is the angle
 * between the velocity vector and the local horizontal.
 *
 * Ref: Vallado Sec.2.3.1
 * ================================================================ */

void velocity_components(double a, double e, double nu, double mu,
                          double *v_r, double *v_theta) {
    double p = a * (1.0 - e * e);
    double h_mag = sqrt(mu * p);
    double r = radial_distance_elliptic(a, e, nu);

    *v_r = mu / h_mag * e * sin(nu);
    *v_theta = h_mag / r;
}

double flight_path_angle(double a, double e, double nu) {
    double v_r, v_theta;
    velocity_components(a, e, nu, 1.0, &v_r, &v_theta);
    if (fabs(v_theta) < 1e-15) {
        return (v_r > 0 ? HALFPI : -HALFPI);
    }
    return atan2(v_r, v_theta);
}

/* ================================================================
 * L3: Laplace-Runge-Lenz (LRL) Vector
 *
 * e_vec = (v x h)/mu - r_hat
 *
 * This vector points from the focus to periapsis, and its magnitude
 * equals the eccentricity. It is a constant of motion specific to
 * the 1/r potential, reflecting the hidden SO(4) symmetry of the
 * Kepler problem.
 *
 * Ref: Goldstein Sec.3.7, Murray & Dermott Sec.2.8
 * ================================================================ */

Vector3 eccentricity_vector(Vector3 r, Vector3 v, double mu) {
    Vector3 h = vec3_cross(r, v);
    Vector3 v_cross_h = vec3_cross(v, h);
    Vector3 r_hat = vec3_unit(r);
    return vec3_sub(vec3_scale(1.0 / mu, v_cross_h), r_hat);
}

/* ================================================================
 * L3: Derived quantities from state vector
 * ================================================================ */

double semi_major_axis_from_state(Vector3 r, Vector3 v, double mu) {
    double r_mag = vec3_norm(r);
    double v_mag = vec3_norm(v);
    double xi = 2.0 / r_mag - v_mag * v_mag / mu;
    if (fabs(xi) < 1e-30) return 1e30;
    return 1.0 / xi;
}

double eccentricity_from_state(Vector3 r, Vector3 v, double mu) {
    return vec3_norm(eccentricity_vector(r, v, mu));
}

double angular_momentum_magnitude(Vector3 r, Vector3 v) {
    return vec3_norm(vec3_cross(r, v));
}

Vector3 angular_momentum_vector(Vector3 r, Vector3 v) {
    return vec3_cross(r, v);
}

double areal_velocity(Vector3 r, Vector3 v) {
    return angular_momentum_magnitude(r, v) / 2.0;
}

/* ================================================================
 * L2: Orbit Shape Parameters
 * ================================================================ */

double flattening_factor(double a, double e) {
    if (e >= 1.0 || a <= 0.0) return 0.0;
    double r_p = a * (1.0 - e);
    double r_a = a * (1.0 + e);
    if (r_a < 1e-30) return 0.0;
    return (r_a - r_p) / r_a;
}

double semi_minor_axis(double a, double e) {
    if (e >= 1.0) return 0.0;
    return a * sqrt(1.0 - e * e);
}

double impact_parameter_hyperbolic(double a, double e) {
    if (e <= 1.0) return 0.0;
    return fabs(a) * sqrt(e * e - 1.0);
}

double bending_angle_hyperbolic(double e) {
    if (e <= 1.0) return 0.0;
    return 2.0 * asin(1.0 / e);
}

File: mini-celestial-mechanics/src/two_body.jl
# two_body.jl — 二体问题: vis-viva, 轨道传播, 飞行时间
# 参考: Goldstein Ch.3, Murray & Dermott Ch.2

# ============================================================
# Vis-Viva 方程与能量
# ============================================================

"""
Vis-Viva 方程: v² = μ(2/r - 1/a)

参数:
- r: 当前位置半径
- a: 半长径 (双曲线为负)
- mu: 引力参数
"""
vis_viva(r::Float64, a::Float64; mu::Float64=G_SUN) = sqrt(mu * (2.0/r - 1.0/a))

"""
轨道能量: ε = -μ/(2a) (椭圆<0, 抛物线=0, 双曲线>0)
"""
orbital_energy(a::Float64; mu::Float64=G_SUN) = -mu / (2.0 * a)

"""
轨道周期: T = 2π√(a³/μ)
"""
orbital_period(a::Float64; mu::Float64=G_SUN) = 2π * sqrt(a^3 / mu)

"""
平均角速度 (mean motion): n = √(μ/a³) = 2π/T
"""
mean_motion(a::Float64; mu::Float64=G_SUN) = sqrt(mu / a^3)

# ============================================================
# 轨道中的位置/速度函数
# ============================================================

"""
椭圆轨道的径向距离: r(ν) = a(1-e²)/(1+e cos ν)
"""
radial_distance_elliptic(a::Float64, e::Float64, nu::Float64) = a * (1.0 - e^2) / (1.0 + e * cos(nu))

"""
椭圆轨道的位置 (轨道平面内, x 指向近心点)
"""
function position_in_orbital_plane(a::Float64, e::Float64, nu::Float64)
    r_mag = radial_distance_elliptic(a, e, nu)
    return [r_mag * cos(nu), r_mag * sin(nu), 0.0]
end

"""
椭圆轨道的速度 (轨道平面内)
"""
function velocity_in_orbital_plane(a::Float64, e::Float64, nu::Float64; mu::Float64=G_SUN)
    r_mag = radial_distance_elliptic(a, e, nu)
    p = a * (1.0 - e^2)
    h_mag = sqrt(mu * p)
    return [
        -mu / h_mag * sin(nu),
        mu / h_mag * (e + cos(nu)),
        0.0
    ]
end

"""
半通径: p = a(1-e²)，对于抛物线: p = h²/μ
"""
semi_latus_rectum(a::Float64, e::Float64) = a * (1.0 - e^2)

# ============================================================
# 飞行时间
# ============================================================

"""
椭圆弧飞行时间 (Kepler 方程)

Δt = √(a³/μ) · [ (E₂ - e sin E₂) - (E₁ - e sin E₁) ]

参数:
- a, e: 轨道根数
- nu1, nu2: 起点和终点的真近点角
- mu: 引力参数

返回: Δt
"""
function time_of_flight_elliptic(a::Float64, e::Float64, nu1::Float64, nu2::Float64;
                                  mu::Float64=G_SUN)
    E1 = true_to_eccentric_anomaly(nu1, e)
    E2 = true_to_eccentric_anomaly(nu2, e)
    M1 = eccentric_to_mean_anomaly(E1, e)
    M2 = eccentric_to_mean_anomaly(E2, e)
    n = mean_motion(a; mu=mu)
    delta_M = mod(M2 - M1 + 2π, 2π)
    return delta_M / n
end

"""
从平近点角计算经过的时间

t = M / n (从近心点开始)
"""
function time_since_periapsis(a::Float64, e::Float64, nu::Float64; mu::Float64=G_SUN)
    E = true_to_eccentric_anomaly(nu, e)
    M = eccentric_to_mean_anomaly(E, e)
    n = mean_motion(a; mu=mu)
    return M / n
end

"""
给定经过的时间，计算真近点角 (Kepler 传播)

传播初始真近点角 ν₀ 到 Δt 之后
"""
function propagate_anomaly(a::Float64, e::Float64, nu0::Float64, delta_t::Float64;
                            mu::Float64=G_SUN)
    E0 = true_to_eccentric_anomaly(nu0, e)
    M0 = eccentric_to_mean_anomaly(E0, e)
    n = mean_motion(a; mu=mu)
    M = M0 + n * delta_t
    M = mod(M, 2π)
    E = solve_kepler(M, e)
    return eccentric_to_true_anomaly(E, e)
end

# ============================================================
# 近心点/远心点
# ============================================================

"近心距: r_p = a(1-e)"
periapsis_distance(a::Float64, e::Float64) = a * (1.0 - e)

"远心距: r_a = a(1+e) (仅 e<1)"
apoapsis_distance(a::Float64, e::Float64) = a * (1.0 + e)

"近心点速度"
periapsis_velocity(a::Float64, e::Float64; mu::Float64=G_SUN) = sqrt(mu / a * (1.0 + e) / (1.0 - e))

"远心点速度"
apoapsis_velocity(a::Float64, e::Float64; mu::Float64=G_SUN) = sqrt(mu / a * (1.0 - e) / (1.0 + e))

"逃逸速度: v_esc = √(2μ/r)"
escape_velocity(r::Float64; mu::Float64=G_SUN) = sqrt(2.0 * mu / r)

"圆轨道速度: v_circ = √(μ/r)"
circular_velocity(r::Float64; mu::Float64=G_SUN) = sqrt(mu / r)

"""
轨道速度分解 (径向、横向)

v_r = μ/(h) e sin ν    (径向)
v_θ = μ/(h) (1 + e cos ν) = h/r  (横向)
"""
function velocity_components(a::Float64, e::Float64, nu::Float64; mu::Float64=G_SUN)
    p = semi_latus_rectum(a, e)
    h_mag = sqrt(mu * p)
    r_mag = radial_distance_elliptic(a, e, nu)
    v_r = mu / h_mag * e * sin(nu)
    v_theta = h_mag / r_mag
    return (v_r=v_r, v_theta=v_theta)
end

# ============================================================
# 轨道特性
# ============================================================

"""
轨道离心率矢量 Laplace-Runge-Lenz: e = (v×h)/μ - r̂
"""
function eccentricity_vector(r::Vector{Float64}, v::Vector{Float64}; mu::Float64=G_SUN)
    h = cross(r, v)
    return cross(v, h) / mu - r / norm(r)
end

"""
轨道半长径 (从状态向量): a = 1/(2/r - v²/μ)
"""
function semi_major_axis_from_state(r::Vector{Float64}, v::Vector{Float64}; mu::Float64=G_SUN)
    r_mag = norm(r)
    v_mag = norm(v)
    return 1.0 / (2.0/r_mag - v_mag^2/mu)
end

"""
轨道离心率 (从状态向量)
"""
function eccentricity_from_state(r::Vector{Float64}, v::Vector{Float64}; mu::Float64=G_SUN)
    return norm(eccentricity_vector(r, v; mu=mu))
end

export vis_viva, orbital_energy, orbital_period, mean_motion
export radial_distance_elliptic, position_in_orbital_plane, velocity_in_orbital_plane
export semi_latus_rectum
export time_of_flight_elliptic, time_since_periapsis, propagate_anomaly
export periapsis_distance, apoapsis_distance, periapsis_velocity, apoapsis_velocity
export escape_velocity, circular_velocity, velocity_components
export eccentricity_vector, semi_major_axis_from_state, eccentricity_from_state
